processor/src/test/java/org/mapstruct/ap/test/java8stream/StreamMappingTest.java

```diff
@@ -0,0 +1,244 @@
+    }
+
+    @Test
+    public void shouldReverseMapList() {
+        Target target = new Target();
+        target.setStringList( Arrays.asList( "Bob", "Alice" ) );
+
+        Source source = SourceTargetMapper.INSTANCE.targetToSource( target );
+
+        assertThat( source ).isNotNull();
+        assertThat( source.getStringList() ).containsExactly( "Bob", "Alice" );
+    }
+
+    @Test
+    public void shouldMapArrayList() {
+        Source source = new Source();
+        source.setStringArrayList( new ArrayList<String>( Arrays.asList( "Bob", "Alice" ) ).stream() );
+
+        Target target = SourceTargetMapper.INSTANCE.sourceToTarget( source );
+
+        assertThat( target ).isNotNull();
+        assertThat( target.getStringArrayList() ).containsExactly( "Bob", "Alice" );
+    }
+
+    @Test
+    public void shouldReverseMapArrayList() {
+        Target target = new Target();
+        target.setStringArrayList( new ArrayList<String>( Arrays.asList( "Bob", "Alice" ) ) );
+
+        Source source = SourceTargetMapper.INSTANCE.targetToSource( target );
+
+        assertThat( source ).isNotNull();
+        assertThat( source.getStringArrayList() ).containsExactly( "Bob", "Alice" );
+    }
+
+    @Test
+    public void shouldMapSet() {
+        Source source = new Source();
+        source.setStringSet( new HashSet<String>( Arrays.asList( "Bob", "Alice" ) ).stream() );
+
+        Target target = SourceTargetMapper.INSTANCE.sourceToTarget( source );
+
+        assertThat( target ).isNotNull();
+        assertThat( target.getStringSet() ).contains( "Bob", "Alice" );
+    }
+
+    @Test
+    public void shouldReverseMapSet() {
+        Target target = new Target();
+        target.setStringSet( new HashSet<String>( Arrays.asList( "Bob", "Alice" ) ) );
+
+        Source source = SourceTargetMapper.INSTANCE.targetToSource( target );
+
+        assertThat( source ).isNotNull();
+        assertThat( source.getStringSet() ).contains( "Bob", "Alice" );
+    }
+
+    @Test
+    public void shouldMapListToCollection() {
+        Source source = new Source();
+        source.setIntegerList( Arrays.asList( 1, 2 ).stream() );
+
+        Target target = SourceTargetMapper.INSTANCE.sourceToTarget( source );
+
+        assertThat( target ).isNotNull();
+        assertThat( target.getIntegerCollection() ).containsOnly( 1, 2 );
+    }
+
+    @Test
+    public void shouldReverseMapListToCollection() {
+        Target target = new Target();
+        target.setIntegerCollection( Arrays.asList( 1, 2 ) );
+
+        Source source = SourceTargetMapper.INSTANCE.targetToSource( target );
+
+        assertThat( source ).isNotNull();
+        assertThat( source.getIntegerList() ).containsOnly( 1, 2 );
+    }
+
+    @Test
+    public void shouldMapIntegerSetToStringSet() {
+        Source source = new Source();
+        source.setAnotherIntegerSet( new HashSet<Integer>( Arrays.asList( 1, 2 ) ).stream() );
```

Comment:
Instead of generating

    protected Set<String> integerStreamToStringSet(Stream<Integer> stream) {
        if ( stream == null ) {
            return null;
        }

        Set<String> set = new HashSet<String>();

        set.addAll( stream.map( integer -> String.valueOf( integer ) )
            .collect( Collectors.toCollection( HashSet<String>::new ) )
        );

        return set;
    }

How about this:

    protected Set<String> integerStreamToStringSet(Stream<Integer> stream) {
        if ( stream == null ) {
            return null;
        }

        return stream.map( integer -> String.valueOf( integer ) )
            .collect( Collectors.toCollection( HashSet<String>::new ) )
        );
    }

I.e. avoid the adding to another collection.